### PR TITLE
Saga tests for interbit-ui-tools

### PR DESCRIPTION
**Pull request by djlmills**

### CODE REVIEW

File: packages/interbit-ui-tools/src/middleware/browserContext/index.js [hunk 1]
```diff
@@ -3,8 +3,18 @@ const getHtmlConfig = require('./staticHtml')
 
 const getConfig = async () => getHtmlConfig(document)
 
+const getDefaultPort = () => {
```
Comment (btlewan) [inline packages/interbit-ui-tools/src/middleware/browserContext/index.js]: Please add source comments for all new/modified functions.

File: packages/interbit-ui-tools/src/tests/middleware/sagas.test.js [hunk 2]
```diff
@@ -0,0 +1,588 @@
...
+      .not.put.like({
+        action: {
+          type: actionTypes.CHAIN_ERROR
+        }
+      })
+      .run()
+  })
+
+  it('If loading interbit throws, dispatch INTERBIT_ERROR action', async () => {
+    const errorMessage = 'interbit is unavailable today'
```
Comment (nicholeous) [inline packages/interbit-ui-tools/src/tests/middleware/sagas.test.js]: :joy:

File: packages/interbit-ui-tools/src/tests/middleware/sagas.test.js [hunk 3]
```diff
@@ -0,0 +1,588 @@
...
+    const { rootSaga } = middlewareSagas(contextWithoutStaticChains)
+    await expectSaga(rootSaga)
+      .withReducer(reducers)
+      .put.like({
+        action: {
+          type: actionTypes.INITIAL_CONFIG,
+          payload: configWithoutStaticChains
+        }
+      })
+      .put.like({
```
Comment (nicholeous) [inline packages/interbit-ui-tools/src/tests/middleware/sagas.test.js]: Would it be reasonable to also specify `.not.put.like({...CHAIN_LOADING})` here or will this attempt to load a private chain?

File: packages/interbit-ui-tools/src/tests/middleware/sagas.test.js [hunk 4]
```diff
@@ -0,0 +1,588 @@
...
+
+  it('If private chain ID in local storage, load the chain', async () => {
+    const contextWithChainIdInLocalStorage = mockContext({
+      dataStoreContent: privateChainIdInLocalStorage
+    })
+    const { rootSaga } = middlewareSagas(contextWithChainIdInLocalStorage)
+    await expectSaga(rootSaga)
+      .withReducer(reducers)
+      .dispatch(
+        // We need the public chain ID
```
Comment (nicholeous) [inline packages/interbit-ui-tools/src/tests/middleware/sagas.test.js]: Comment provides no context beyond the dispatch call. Why do we need the chain ID?

File: packages/interbit-ui-tools/src/tests/middleware/sagas.test.js [hunk 5]
```diff
@@ -0,0 +1,588 @@
...
+          }
+        }
+      })
+      .put.like({
+        action: {
+          type: actionTypes.CHAIN_BLOCKING,
+          payload: { chainAlias: PRIVATE_CHAIN_ALIAS }
+        }
+      })
+      .not.put.like({
```
Comment (nicholeous) [inline packages/interbit-ui-tools/src/tests/middleware/sagas.test.js]: I really really like these `not.put.likes` :)

File: packages/interbit-ui-tools/src/tests/middleware/sagas.test.js [hunk 6]
```diff
@@ -0,0 +1,588 @@
...
+      })
+      .run()
+  })
+
+  it('If private chain ID not known, sponsor a new chain', async () => {
+    const { rootSaga } = middlewareSagas(mockContext())
+    await expectSaga(rootSaga)
+      .withReducer(reducers)
+      .dispatch(
+        // We need the public chain ID and sponsor config
```
Comment (nicholeous) [inline packages/interbit-ui-tools/src/tests/middleware/sagas.test.js]: Why? This only tells me as much as the dispatch does

File: packages/interbit-ui-tools/src/tests/middleware/sagas.test.js [hunk 7]
```diff
@@ -0,0 +1,588 @@
...
+  it('Switch private chains for second device support', async () => {
+    const OAUTH_PRIVATE_CHAIN_ID = 'e2329e3592040f87a60323423h23g4234...'
+    const contextWithChainIdInLocalStorage = mockContext({
+      dataStoreContent: privateChainIdInLocalStorage
+    })
+    const { rootSaga } = middlewareSagas(contextWithChainIdInLocalStorage)
+    await expectSaga(rootSaga)
+      .withReducer(reducers)
+      .dispatch(
+        // We need the public chain ID
```
Comment (nicholeous) [inline packages/interbit-ui-tools/src/tests/middleware/sagas.test.js]: Why?

File: packages/interbit-ui-tools/src/tests/middleware/sagas.test.js [hunk 8]
```diff
@@ -0,0 +1,588 @@
...
+    const SPONSORED_CHAIN_ALIAS = 'application'
+    const SPONSORED_CHAIN_ID = '920044625fa22243b329e353898cc55fc...'
+    const SPONSORED_CHAIN_COVENANT_HASH = 'fa22243b3290044625e35923898cc55fc...'
+    const { rootSaga } = middlewareSagas(
+      mockContext({ genesisBlockHash: SPONSORED_CHAIN_ID })
+    )
+    await expectSaga(rootSaga)
+      .withReducer(reducers)
+      .dispatch(
+        // We need the public chain ID and sponsor config
```
Comment (nicholeous) [inline packages/interbit-ui-tools/src/tests/middleware/sagas.test.js]: Why?

File: packages/interbit-ui-tools/src/tests/testContext/testContext.test.js [hunk 9]
```diff
@@ -64,6 +64,9 @@ describe('testContext', () => {
           assert.ok(interbitContext.publicKey)
           assert.ok(interbitContext.chains)
           assert.ok(interbitContext.localDataStore)
+
+          const reloadedContext = await testContext.waitForInterbit()
+          assert.strictEqual(reloadedContext, interbitContext)
```
Comment (nicholeous) [inline packages/interbit-ui-tools/src/tests/testContext/testContext.test.js]: :+1: Strict :)

File: packages/interbit-ui-tools/src/tests/middleware/sagas.test.js [hunk 10]
```diff
@@ -0,0 +1,588 @@
...
+          type: actionTypes.CHAIN_BLOCKING,
+          payload: { chainAlias: SPONSORED_CHAIN_ALIAS }
+        }
+      })
+      .not.put.like({
+        action: {
+          type: actionTypes.CHAIN_ERROR
+        }
+      })
+      .run()
```
Comment (nicholeous) [inline packages/interbit-ui-tools/src/tests/middleware/sagas.test.js]: :+1: very thorough

File: packages/interbit-ui-tools/src/middleware/browserContext/index.js (hunk 1 above)
Comment (djlmills) [reply packages/interbit-ui-tools/src/middleware/browserContext/index.js]: This is not exposed in the public API and source comments are not applicable.

File: packages/interbit-ui-tools/src/tests/middleware/sagas.test.js (hunk 3 above)
Comment (djlmills) [reply packages/interbit-ui-tools/src/tests/middleware/sagas.test.js]: `.not.put.like()` is an assertion about the behaviour and has no side effects. So this makes sense.

Comment (djlmills) [reply packages/interbit-ui-tools/src/tests/middleware/sagas.test.js]: Done

**Pull request merged**